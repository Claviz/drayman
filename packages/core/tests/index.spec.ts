import { componentInstances, handleComponentEvent, saveComponent, handleEventHubEvent, onDestroyComponentInstance, onInitializeComponentInstance, getElementsScriptPaths } from '../dist';
import fs from 'fs-extra';
import path from 'path';

describe('', () => {

    const onError = jest.fn();
    const onSuccess = jest.fn();

    beforeAll(async () => {
        const script = await fs.readFile('./tests/components/buttons.tsx', 'utf-8');
        await saveComponent({ script, outputFile: './tests/dist/components/buttons.js' });
    });

    beforeEach(() => {
        jest.resetAllMocks();
    });

    test('elements paths', async () => {
        const elementPaths = await getElementsScriptPaths({ nodeModulesPath: path.join(__dirname, '../../framework/node_modules') });
        const pathsWithRelativeRemoved = {};
        for (const element of Object.keys(elementPaths)) {
            pathsWithRelativeRemoved[element] = path.relative(__dirname, elementPaths[element]);
        }
        expect(pathsWithRelativeRemoved).toMatchSnapshot();
    });

    test('component user interaction', async () => {
        const consoleMessages = [];
        const messages = await (() => new Promise<{ type, payload, componentInstanceId }[]>((resolve, reject) => {
            const componentInstanceId = 'instance-1';
            const messages = [];
            onInitializeComponentInstance({
                browserCommands: [],
                onComponentInstanceConsole: ({ text }) => { consoleMessages.push(text) },
                componentInstanceId,
                componentName: 'buttons',
                componentRootDir: 'tests/dist/components',
                connectionId: 'connection-1',
                componentOptions: { text: 'Hello, world!' },
                emit: async (message) => {
                    messages.push(message);
                    switch (messages.length) {
                        // received initial view
                        case 1: {
                            handleComponentEvent({
                                componentInstanceId,
                                eventName: '/0/button/click',
                                files: [],
                                options: {},
                                onError,
                                onSuccess: () => {
                                    onDestroyComponentInstance({ componentInstanceId: componentInstanceId });
                                },
                            });
                            return;
                        }
                        // received view after button click
                        case 2: {
                            return;
                        }
                        // received message about component destroy
                        case 3: {
                            expect(componentInstances[componentInstanceId]).not.toBeDefined();
                            resolve(messages);
                        }
                    }
                },
            });
        }))();
        expect(messages).toMatchSnapshot();
        expect(consoleMessages).toMatchSnapshot();
    });

    test('trying to perform event on non-existing component instance result in an error', async () => {
        const messages = await (() => new Promise<{ type, payload, componentInstanceId }[]>((resolve, reject) => {
            const componentInstanceId = 'instance-2';
            const messages = [];
            onInitializeComponentInstance({
                browserCommands: [],
                onComponentInstanceConsole: ({ text }) => { console.log(text) },
                componentInstanceId,
                componentName: 'buttons',
                componentRootDir: 'tests/dist/components',
                connectionId: 'connection-2',
                componentOptions: { text: 'Hello, world!' },
                emit: async (message) => {
                    messages.push(message);
                    switch (messages.length) {
                        // received initial view
                        case 1: {
                            handleComponentEvent({
                                componentInstanceId: 'non-existing-instance',
                                eventName: '/0/button/onClick',
                                files: [],
                                options: {},
                                onError: () => {
                                    onDestroyComponentInstance({ componentInstanceId: componentInstanceId });
                                },
                                onSuccess,
                            });
                            return;
                        }
                        // received message about component destroy
                        case 2: {
                            expect(componentInstances[componentInstanceId]).not.toBeDefined();
                            resolve(messages);
                        }
                    }
                },
            });
        }))();
        expect(messages).toMatchSnapshot();
    });

    // test('navigating', async () => {
    //     const messages = await (() => new Promise<{ type, payload, componentInstanceId }[]>((resolve, reject) => {
    //         const componentInstanceId = 'instance-1';
    //         const messages = [];
    //         onInitializeComponentInstance({
    //             browserCommands: [],
    //             onComponentInstanceConsole: ({ text }) => { console.log(text) },
    //             componentInstanceId,
    //             componentName: 'buttons',
    //             componentRootDir: 'tests/dist/components',
    //             connectionId: 'connection-1',
    //             componentOptions: { text: 'Hello, world!' },
    //             emit: async (message) => {
    //                 messages.push(message);
    //                 switch (messages.length) {
    //                     case 1: {
    //                         handleComponentEvent({
    //                             componentInstanceId,
    //                             eventName: '/3/button/onClick',
    //                             files: [],
    //                             options: {},
    //                             onError,
    //                             onSuccess: () => {
    //                                 onDestroyComponentInstance({ componentInstanceId: componentInstanceId });
    //                             },
    //                         });
    //                         return;
    //                     }
    //                     case 2: {
    //                         return;
    //                     }
    //                     case 3: {
    //                         resolve(messages);
    //                     }
    //                 }
    //             },
    //         });
    //     }))();
    //     expect(messages).toMatchSnapshot();
    // });

    // test('should change location for specific connection', async () => {
    //     const messages = await (() => new Promise<{ type, payload, componentInstanceId }[]>((resolve, reject) => {
    //         const componentInstanceId = 'instance-1';
    //         const messages = [];
    //         onInitializeComponentInstance({
    //             browserCommands: [],
    //             onComponentInstanceConsole: ({ text }) => { console.log(text) },
    //             componentInstanceId,
    //             componentName: 'buttons',
    //             componentRootDir: 'tests/dist/components',
    //             connectionId: 'connection-1',
    //             componentOptions: { text: 'Hello, world!' },
    //             emit: async (message) => {
    //                 messages.push(message);
    //                 switch (messages.length) {
    //                     case 1: {
    //                         onLocationChange({
    //                             connectionId: 'connection-1',
    //                             location: { href: 'http://localhost/login' },
    //                         });
    //                         return;
    //                     }
    //                     case 2: {
    //                         onDestroyComponentInstance({ componentInstanceId: componentInstanceId });
    //                         return;
    //                     }
    //                     case 3: {
    //                         resolve(messages);
    //                     }
    //                 }
    //             },
    //         });
    //     }))();
    //     expect(messages).toMatchSnapshot();
    // });

    // test('should not change location for other connection', async () => {
    //     const send = jest.fn();
    //     const messages = await (() => new Promise<{ type, payload, componentInstanceId }[]>((resolve, reject) => {
    //         const componentInstanceId = 'instance-1';
    //         const messages = [];
    //         onInitializeComponentInstance({
    //             browserCommands: [],
    //             onComponentInstanceConsole: ({ text }) => { console.log(text) },
    //             componentInstanceId,
    //             componentName: 'buttons',
    //             componentRootDir: 'tests/dist/components',
    //             connectionId: 'connection-1',
    //             componentOptions: { text: 'Hello, world!' },
    //             emit: async (message) => {
    //                 messages.push(message);
    //                 switch (messages.length) {
    //                     case 1: {
    //                         componentInstances[componentInstanceId].worker.handleLocationChange = send;
    //                         onLocationChange({
    //                             connectionId: 'connection-2',
    //                             location: { href: 'http://localhost/login' },
    //                         });
    //                         onDestroyComponentInstance({ componentInstanceId: componentInstanceId });
    //                         return;
    //                     }
    //                     case 2: {
    //                         resolve(messages);
    //                     }
    //                 }
    //             },
    //         });
    //     }))();
    //     expect(messages).toMatchSnapshot();
    //     expect(send.mock.calls.length).toEqual(0);
    //     expect(onError.mock.calls.length).toEqual(0);
    //     expect(onSuccess.mock.calls.length).toEqual(0);
    // });
});
