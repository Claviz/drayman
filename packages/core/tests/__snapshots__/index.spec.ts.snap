// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component user interaction 1`] = `
Array [
  Object {
    "componentInstanceId": "instance-1",
    "payload": Object {
      "updateId": 1,
      "view": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/button/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {
            "on": Object {
              "click": true,
            },
          },
          "key": "/0/button",
          "sel": "button",
        },
        Object {
          "children": Array [],
          "data": Object {},
          "key": "/1/h1",
          "sel": "h1",
        },
      ],
    },
    "type": "view",
  },
  Object {
    "componentInstanceId": "instance-1",
    "payload": Object {
      "updateId": 2,
      "view": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/button/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {
            "on": Object {
              "click": true,
            },
          },
          "key": "/0/button",
          "sel": "button",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/1/h1/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {},
          "key": "/1/h1",
          "sel": "h1",
        },
      ],
    },
    "type": "view",
  },
  Object {
    "componentInstanceId": "instance-1",
    "payload": Object {},
    "type": "componentInstanceDestroyed",
  },
]
`;

exports[`component user interaction 2`] = `
Array [
  "ComponentInstance object has id: true
",
  "You've clicked a button!
",
  "destroyed!
",
]
`;

exports[`elements paths 1`] = `
Object {
  "drayman-button": "../../framework/node_modules/@drayman/elements/dist/button/main.js",
  "drayman-checkbox": "../../framework/node_modules/@drayman/elements/dist/checkbox/main.js",
  "drayman-claviz-charts": "../../framework/node_modules/@drayman/elements/dist/claviz-charts/main.js",
  "drayman-datepicker": "../../framework/node_modules/@drayman/elements/dist/datepicker/main.js",
  "drayman-file-uploader": "../../framework/node_modules/@drayman/elements/dist/file-uploader/main.js",
  "drayman-menu": "../../framework/node_modules/@drayman/elements/dist/menu/main.js",
  "drayman-nebula": "../../framework/node_modules/@drayman/elements/dist/nebula/main.js",
  "drayman-ngx-charts": "../../framework/node_modules/@drayman/elements/dist/ngx-charts/main.js",
  "drayman-ngx-graph": "../../framework/node_modules/@drayman/elements/dist/ngx-graph/main.js",
  "drayman-number-field": "../../framework/node_modules/@drayman/elements/dist/number-field/main.js",
  "drayman-pdf-viewer": "../../framework/node_modules/@drayman/elements/dist/pdf-viewer/main.js",
  "drayman-radio-group": "../../framework/node_modules/@drayman/elements/dist/radio-group/main.js",
  "drayman-select": "../../framework/node_modules/@drayman/elements/dist/select/main.js",
  "drayman-table": "../../framework/node_modules/@drayman/elements/dist/table/main.js",
  "drayman-text-field": "../../framework/node_modules/@drayman/elements/dist/text-field/main.js",
  "drayman-textarea-field": "../../framework/node_modules/@drayman/elements/dist/textarea-field/main.js",
  "drayman-timepicker": "../../framework/node_modules/@drayman/elements/dist/timepicker/main.js",
  "drayman-youtube-player": "../../framework/node_modules/@drayman/elements/dist/youtube-player/main.js",
}
`;

exports[`trying to perform event on non-existing component instance result in an error 1`] = `
Array [
  Object {
    "componentInstanceId": "instance-2",
    "payload": Object {
      "updateId": 1,
      "view": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "/0/button/0/text",
              "text": "Hello, world!",
            },
          ],
          "data": Object {
            "on": Object {
              "click": true,
            },
          },
          "key": "/0/button",
          "sel": "button",
        },
        Object {
          "children": Array [],
          "data": Object {},
          "key": "/1/h1",
          "sel": "h1",
        },
      ],
    },
    "type": "view",
  },
  Object {
    "componentInstanceId": "instance-2",
    "payload": Object {},
    "type": "componentInstanceDestroyed",
  },
]
`;
